hive/src/main/java/com/netflix/iceberg/hive/HiveTableOperations.java

```diff
@@ -0,0 +1,224 @@
+
+  private final IMetaStoreClient metaStoreClient;
+  private final String database;
+  private final String tableName;
+
+  protected HiveTableOperations(Configuration conf, IMetaStoreClient metaStoreClient, String database, String table) {
+    super(conf);
+    this.metaStoreClient = metaStoreClient;
+    this.database = database;
+    tableName = table;
+    refresh();
+  }
+
+  @Override
+  public TableMetadata refresh() {
+    try {
+      final Table table = metaStoreClient.getTable(database, tableName);
+      String tableType = table.getParameters().get(TABLE_TYPE_PROP);
+
+      if (tableType == null || !tableType.equalsIgnoreCase(ICEBERG_TABLE_TYPE_VALUE)) {
+        throw new IllegalArgumentException(format("Invalid tableName, not Iceberg: %s.%s", database, table));
+      }
+
+      String metadataLocation = table.getParameters().get(METADATA_LOCATION_PROP);
+      if(metadataLocation == null) {
+        throw new IllegalArgumentException(format("missing metadata_location: %s.%s", database, tableName));
+      }
+
+      refreshFromMetadataLocation(metadataLocation);
+    } catch (NoSuchObjectException e) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(format("No such table: %s.%s", database, tableName));
+      }
+    } catch (TException e) {
+      throw new RuntimeException(e);
+    }
+    return current();
+  }
+
+  @Override
+  public void commit(TableMetadata base, TableMetadata metadata) {
+    // if the metadata is already out of date, reject it
+    if (base != current()) {
+      throw new CommitFailedException("Cannot commit changes based on stale tableName metadata");
+    }
+
+    // if the metadata is not changed, return early
+    if (base == metadata) {
+      LOG.info("Nothing to commit.");
+      return;
+    }
+
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+
+    boolean threw = true;
+    Optional<Long> lockId = Optional.empty();
+    try {
+      lockId = acquireLock();
+      // TODO add lock heart beating for cases where default lock timeout is too low.
+      Table tbl;
+      if (base != null) {
+        tbl = metaStoreClient.getTable(database, tableName);
+      } else {
+        final long currentTimeMillis = System.currentTimeMillis();
+        final StorageDescriptor storageDescriptor = storageDescriptor(metadata.schema());
+        tbl = new Table(tableName,
+                database,
+                System.getProperty("user.name"),
+                (int) currentTimeMillis / 1000,
+                (int) currentTimeMillis / 1000,
+                Integer.MAX_VALUE,
+                storageDescriptor,
+                Collections.emptyList(),
+                new HashMap<>(),
+                null,
+                null,
+                ICEBERG_TABLE_TYPE_VALUE);
+      }
+
+      // unnecessary but setting in case tableName.getParameters() starts returning a deep copy.
+      setParameters(newMetadataLocation, tbl);
```

Comment:
This also needs to check that the table's `metadata_location` that was just fetched while holding the lock is equal to `currentMetadataLocation`. If that doesn't happen, then this isn't an atomic swap and two concurrent writers that start at the same time could corrupt the table because the last one wins.

Reply:
That was plain dumb. Sorry about that. Added the check and also added a test to verify the behavior.